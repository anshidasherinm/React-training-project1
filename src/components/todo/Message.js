import React from "react";
import { useAppSelector } from "store/redux-hooks";
import { LinearProgress } from "@mui/material";
import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles({
  message: {
    color: "white",
    width: "100%",
  },

  errorMessage: {
    color: "red",
    width: "80%",
  },
});

const styles = {
  progress: {
    backgroundColor: "white",
    "& .MuiLinearProgress-barColorPrimary": {
      //
      backgroundColor: "orange",
    },
  },
};
function Message() {
  const classes = useStyles();
  const {
    addingTodoItem,
    loadingList,
    updatingTodo,
    deletingTodoItem,
    addingTodoItemError,
    loadingListError,
    updatingTodoError,
    deletingTodoItemError,
    // } = useSelector((state) => state.todo);
  } = useAppSelector((state) => state.todo);
  if (updatingTodo) {
    return (
      <div className={classes.message}>
        <LinearProgress sx={styles.progress} />
      </div>
    );
  }

  if (addingTodoItem) {
    return (
      <div className={classes.message}>
        <LinearProgress sx={styles.progress} />
      </div>
    );
  }

  if (loadingList) {
    return <div className={classes.message}>Loading list...</div>;
  }

  if (deletingTodoItem) {
    return (
      <div className={classes.message}>
        <LinearProgress sx={styles.progress} />
      </div>
    );
  }

  if (
    addingTodoItemError ||
    loadingListError ||
    updatingTodoError ||
    deletingTodoItemError
  ) {
    return (
      <div className={classes.errorMessage}>
        Error ! could not perform action
      </div>
    );
  }

  return null;
}

export default Message;
