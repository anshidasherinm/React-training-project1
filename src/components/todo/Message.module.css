.message {
  color: white;
  /* margin-top: 1rem; */
  width: 100%;
}

.errorMessage {
  color: red;
  width: 80%;
}
