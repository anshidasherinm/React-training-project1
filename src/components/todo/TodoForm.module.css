.errormsg {
  height: 2rem;
  font-size: medium;
  color: red;
  width: 80%;
  /* background-color: yellow; */
  margin-left: 5rem;
}
.todoForm {
  padding: 0rem;
  height: 100%;
  width: 90%;
  display: flex;
  flex-direction: column;
  font-size: large;
  align-items: center;
  justify-content: center;
}

.labels {
  width: 25%;
  margin-left: 1rem;
  margin-right: 1rem;
}

.message {
  margin-top: 1rem;
  margin-bottom: 1rem;
  /* height: 2rem; */
  width: 100%;
  /* background-color: red; */
  display: flex;
  align-items: center;
  justify-content: center;
}
