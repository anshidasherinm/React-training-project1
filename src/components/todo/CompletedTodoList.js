import React from "react";
import { useAppSelector } from "store/redux-hooks";
import TodoItem from "components/todo/TodoItem";
import { Card, List } from "@mui/material";

const CompletedTodoList = (props) => {
  const styles = {
    card_empty: {
      display: "flex",
      flexDirection: "column",
      justifyContent: "center",
      alignItems: "center",
      overflowY: "auto",
      width: "100%",
      color: "white",
      backgroundColor: "#383838",
    },
    card_notempty: {
      display: "flex",
      flexDirection: "column",
      justifyContent: "center",
      alignItems: "center",
      width: "100%",
      backgroundColor: "rgb(42, 4, 77)",
    },
    list: {
      width: "100%",
      bgcolor: "background.paper",
      padding: "0px",
      backgroundColor: "#383838",
      overflowY: "auto",
      display: "flex",
      flexDirection: "column",
      height: "100%",
      alignItems: "center",
    },
  };

  const todoItems = useAppSelector((state) =>
    state.todo.items.filter((todo) => todo.checked === true)
  );

  if (todoItems.length === 0) {
    return <Card style={styles.card_empty}>No completed todos</Card>;
  }

  return (
    <Card style={styles.card_notempty}>
      <List sx={styles.list} data-testid="todo-list">
        {todoItems.map((todo, index) => {
          const labelId = `checkbox-list-label-${todo.id}`;

          return (
            <TodoItem
              data-testid={`item-${index}`}
              id={todo.key}
              title={todo.title}
              key={todo.key}
              check={todo.checked}
              labelId={labelId}
            />
          );
        })}
      </List>
    </Card>
  );
};

export default CompletedTodoList;
