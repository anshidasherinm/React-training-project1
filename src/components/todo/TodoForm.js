import React, { useState } from "react";
import { addTodoItem } from "store/todo/todo-thunk";
import { todoActions } from "store/todo/todo-slice";
import { useAppDispatch, useAppSelector } from "store/redux-hooks";
import Message from "components/todo/Message";
import { Button, Card } from "@mui/material";
import { MenuItem, Select, TextField, Grid } from "@mui/material";
import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles({
  errormsg: {
    height: "2rem",
    fontSize: "medium",
    color: "red",
    width: "80%",
    marginLeft: "5rem",
  },
  todoForm: {
    padding: "0rem",
    height: "100%",
    width: "90%",
    display: "flex",
    flexDirection: "column",
    fontSize: "large",
    alignItems: "center",
    justifyContent: "center",
  },

  labels: {
    width: "25%",
    marginLeft: "1rem",
    marginRight: "1rem",
  },

  message: {
    marginTop: "1rem",
    marginBottom: "1rem",
    width: "100%",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
});

const TodoForm = (props) => {
  const dispatch = useAppDispatch();
  const classes = useStyles();
  const labelItems = useAppSelector((state) => state.label.labels);
  const [selectedLabelKey, setSelectedLabelKey] = useState("none");
  const [notValid, setNotValid] = useState(false);
  const [todoItem, setTodoItem] = useState("");

  const formSubmitHandler = async (event) => {
    // event.nativeEvent.stopImmediatePropagation();
    console.log("todo form submitted");
    // event.stopPropagation();
    if (todoItem) {
      const newTodo = {
        id: Math.trunc(Math.random() * 2000),
        title: todoItem,
        labelKey: selectedLabelKey,
        checked: false,
      };
      dispatch(todoActions.setBacktoAllItems());
      dispatch(addTodoItem(newTodo));
      setNotValid(false);
      setTodoItem("");
    } else {
      setNotValid(true);
    }
  };
  const getLabelValue = (e) => {
    setSelectedLabelKey(e.target.value);
  };

  const styles = {
    card: {
      display: "flex",
      flexDirection: "column",
      alignItems: "center",
      justifyContent: "center",
      backgroundColor: "#383838",
      paddingTop: "1rem",
      marginBottom: "1rem",
    },
    grid_parent: {
      width: "100%",
      display: "flex",
      flexDirection: "row",
      alignItems: "center",
      justifyContent: "center",
    },
    grid_child1: {
      display: "flex",
      marginTop: "1rem",
      flexDirection: "row",
      alignItems: "center",

      justifyContent: "space-evenly",
    },
    textfield: {
      width: "50%",
      backgroundColor: "#494848be",
      color: "black",
      ".MuiOutlinedInput-notchedOutline": {
        borderColor: "orange !important",
        color: "orange",
      },
      ".MuiInputBase-input": {
        color: "white",
      },
    },
    select: {
      width: "100%",
      backgroundColor: "#494848be",
      fontSize: "1rem",
      color: "orange",
      ".MuiOutlinedInput-notchedOutline": {
        borderColor: "orange !important",
      },
      "& .MuiSvgIcon-root": {
        color: "orange",
      },
    },
    grid_child2: {
      display: "flex",
      marginTop: "1rem",
      flexDirection: "row",
      alignItems: "center",
      justifyContent: "center",
    },
    button: {
      backgroundColor: "orange",
      color: "white",
      fontSize: "1.3rem",
      fontWeight: "large",
      ":hover": {
        backgroundColor: "#ffb347",
        color: "white",
        border: "none",
      },
    },
    inputlabelprops: { fontSize: "1.2rem", color: "orange" },
  };

  return (
    <React.Fragment>
      <Card variant="outlined" style={styles.card}>
        <form className={classes.todoForm}>
          <Grid container spacing={0} style={styles.grid_parent}>
            <Grid item xs={12} md={9} style={styles.grid_child1}>
              <TextField
                sx={styles.textfield}
                InputLabelProps={{
                  style: styles.inputlabelprops,
                }}
                id="todoitem-name"
                label="Add Todo"
                variant="outlined"
                value={todoItem}
                onChange={(todo) => setTodoItem(todo.target.value)}
                data-testid={"input-todo-title"}
              />
              <div className={classes.labels}>
                <Select
                  sx={styles.select}
                  labelId="select-label"
                  id="simple-select"
                  value={selectedLabelKey}
                  onChange={getLabelValue}
                  fullWidth
                  data-testid={"label-select"}
                >
                  <MenuItem value={`none`}>None</MenuItem>
                  {labelItems.map((label, index) => {
                    return (
                      <MenuItem value={`${label.key}`} key={label.key}>
                        {label.name}
                      </MenuItem>
                    );
                  })}
                </Select>
              </div>
            </Grid>
            <Grid item xs={12} md={2} style={styles.grid_child2}>
              <Button
                size="large"
                variant="outlined"
                onClick={formSubmitHandler}
                sx={styles.button}
                data-testid={"todo-add"}
              >
                Add
              </Button>
            </Grid>
          </Grid>
          <div className={classes.message}>
            <Message />
          </div>
          <span className={classes.errormsg}>
            {notValid && "Please add a todo !"}
          </span>
        </form>
      </Card>
    </React.Fragment>
  );
};

export default TodoForm;
