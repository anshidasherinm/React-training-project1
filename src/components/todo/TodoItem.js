import React from "react";
import { updateTodo, deleteTodo } from "store/todo/todo-thunk";
import { useAppDispatch, useAppSelector } from "store/redux-hooks";
import { makeStyles } from "@material-ui/core/styles";
import { Checkbox, ListItemIcon, ListItemText, ListItem } from "@mui/material";

const useStyles = makeStyles({
  remove: {
    color: "red",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    width: "1rem",
    fontSize: "larger",
    marginRight: "1rem",
    "&:hover": {
      color: "rgb(243, 65, 65)",
    },

    checkBoxStyle: {
      color: "white",
      // "&.Mui-checked": classes.muiChecked
      "&.Mui-checked": { color: "orange" },
    },
  },

  muiChecked: {
    color: "orange",
  },
});

const TodoItem = (props) => {
  // const dispatch = useDispatch();
  // const todoItems = useSelector((state) => state.todo.items);
  const dispatch = useAppDispatch();
  const todoItems = useAppSelector((state) => state.todo.items);

  const classes = useStyles();

  const checkBoxClickHandler = (key, e) => {
    if (e.target.checked) {
      const item = todoItems.find((e) => e.key === key);
      dispatch(updateTodo({ ...item, checked: true }));
    } else {
      const item = todoItems.find((e) => e.key === key);

      dispatch(updateTodo({ ...item, checked: false }));
    }
  };

  const removeTodoHandler = (key) => {
    const item = todoItems.find((e) => e.key === key); //expression
    dispatch(deleteTodo({ ...item, checked: true })); //statement
  };

  const styles = {
    //above function
    listItem: {
      paddingLeft: "1rem",
      width: "100%",
      boxSizing: "border-box",
      backgroundColor: "#383838",
      minHeight: "5rem",
      borderBottom: "3px solid #282828",
    },

    checkbox: {
      color: "white",
      // "&.Mui-checked": classes.muiChecked
      "&.Mui-checked": { color: "orange" },
    },

    listItemText: {
      overflowWrap: "normal",
      "& .MuiTypography-root": {
        color: "orange",
        fontSize: "1.1rem",
        fontWeight: "normal",
      },
      marginRight: "2rem",
    },
    listItemIcon1: { minWidth: "40px" },
    listItemIcon2: { minWidth: "40px", marginLeft: "1rem" },
  };

  return (
    <ListItem
      sx={styles.listItem}
      key={props.id}
      disablePadding
      // data-testid={"todo-listItem"}
    >
      <ListItemIcon sx={styles.listItemIcon1}>
        <Checkbox
          // sx={classes.checkBoxStyle}
          // className={classes.checkBoxStyle}
          sx={styles.checkbox}
          edge="start"
          checked={props.check}
          onChange={(e) => checkBoxClickHandler(props.id, e)}
          inputProps={{ "aria-labelledby": props.labelId }}
          data-testid={"todo-checkbox"}
        />
      </ListItemIcon>

      <ListItemText
        id={props.labelId}
        primary={`${props.title}`}
        sx={styles.listItemText}
        data-testid={"todo-listItemtext"}
      />

      <ListItemIcon sx={styles.listItemIcon2} data-testid={"todo-removeicon"}>
        {!props.check ? (
          <i
            className={`fa fa-remove ${classes.remove}`}
            onClick={(e) => removeTodoHandler(props.id)}
          />
        ) : null}
      </ListItemIcon>
    </ListItem>
  );
};

export default TodoItem;
