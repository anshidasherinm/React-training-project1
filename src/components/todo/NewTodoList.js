import React from "react";
import { Card, List } from "@mui/material";
import TodoItem from "components/todo/TodoItem";
import { useAppSelector } from "store/redux-hooks";

const NewTodoList = (props) => {
  const todoItems = useAppSelector((state) =>
    state.todo.items.filter((todo) => todo.checked === false)
  );

  const styles = {
    card_empty: {
      display: "flex",
      flexDirection: "column",
      justifyContent: "center",
      alignItems: "center",
      overflowY: "auto",
      width: "100%",
      color: "white",
      backgroundColor: "#383838",
    },
    card_notempty: {
      display: "flex",
      flexDirection: "column",
      justifyContent: "center",
      alignItems: "center",
      width: "100%",
      backgroundColor: "#383838",
    },
    list: {
      padding: "0px",
      width: "100%",
      bgcolor: "background.paper",
      overflowY: "auto",
      display: "flex",
      flexDirection: "column",
      height: "100%",
      alignItems: "center",
      backgroundColor: "#383838",
    },
  };

  if (todoItems.length === 0) {
    return <Card style={styles.card_empty}>Add a todo</Card>;
  }

  return (
    <Card style={styles.card_notempty}>
      <List sx={styles.list}>
        {todoItems.map((todo) => {
          const labelId = `checkbox-list-label-${todo.id}`;

          return (
            <TodoItem
              id={todo.key}
              title={todo.title}
              key={todo.key}
              check={todo.checked}
              lalabelId={labelId}
            />
          );
        })}
      </List>
    </Card>
  );
};

export default NewTodoList;
