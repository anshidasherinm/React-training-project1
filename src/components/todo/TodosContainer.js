import React, { useEffect } from "react";
import { getTodoList } from "store/todo/todo-thunk";
import { useAppDispatch, useAppSelector } from "store/redux-hooks";
import NewTodoList from "components/todo/NewTodoList";
import CompletedTodoList from "components/todo/CompletedTodoList";
import { Card } from "@mui/material";
import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles({
  seperator: {
    width: "0.5px",
    backgroundColor: "grey",
    height: "100%",
    border: "0.5px solid grey",
  },
  card_style: {
    display: "flex",
    flexDirection: "coloumn",
    justifyContent: "space-between",
    height: "20rem",
    backgroundColor: "#383838",
  },
});

const TodosContainer = (props) => {
  const classes = useStyles();
  const styles = {
    card: {
      display: "flex",
      flexDirection: "coloumn",
      justifyContent: "space-between",
      height: "20rem",
      backgroundColor: "#383838",
    },
  };

  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(getTodoList());
  }, []);

  return (
    <Card variant="outlined" style={styles.card}>
      <NewTodoList />
      <div className={classes.seperator} />
      <CompletedTodoList />
    </Card>
  );
};

export default TodosContainer;
