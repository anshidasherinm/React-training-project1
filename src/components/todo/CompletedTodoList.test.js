import { render, screen } from "@testing-library/react";
import { useAppDispatch, useAppSelector } from "store/redux-hooks";
import { testUseAppSelector } from "store/test-app-selector";
import CompletedTodoList from "./CompletedTodoList";
import "@testing-library/jest-dom";
import { getTodoList } from "store/todo/todo-thunk";
import { shallow } from "enzyme";
import TodoItem from "./TodoItem";

jest.mock("store/redux-hooks");
jest.mock("store/todo/todo-thunk");

describe("CompletedTodoList", () => {
  beforeEach(() => {
    useAppSelector.mockImplementation(testUseAppSelector);
    useAppDispatch.mockReturnValue(jest.fn());
  });
  //   it("renders No completed todos ", () => {
  //     //Arrange
  //     render(<CompletedTodoList />);

  //     //Act
  //     //...nothing

  //     //Assert
  //     const emptyList = screen.getByText(/No completed todos/i);
  //     expect(emptyList).toBeInTheDocument();
  //   });
  it("We show a todo todo in CompletedTodo", async () => {
    render(<CompletedTodoList />);

    const enzymeWrapper = shallow(
      <TodoItem
        id={714}
        key="-NOqX_ZRWmYBbJS0hkJh"
        title="study for exam"
        labelKey="-NOqXGK_kreZB5AW_PFG"
        checked={true}
      />
    );
    expect(enzymeWrapper.text()).toMatchSnapshot(); // 1
  });
});
