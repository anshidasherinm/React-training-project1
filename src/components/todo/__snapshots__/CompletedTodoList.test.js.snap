// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompletedTodoList We show a todo todo in CompletedTodo 1`] = `""`;
